### PR TITLE
Initial implementation of the Break Overlay

### PR DESCRIPTION
**Pull request by EVAST9919**
screenshot-284

### CODE REVIEW

File: osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            InitializeBreaks();
+        }
+
+        public void InitializeBreaks()
+        {
+            if (Breaks != null)
+            {
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs]: Invert this as `if (Breaks == null)`

File: osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                foreach (var b in Breaks)
+                {
+                    if (b.HasEffect)
+                    {
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs]: Invert this as `if (!b.HasEffect)`

**Review by smoogipoo — DISMISSED** [review]
This is looking good code-wise.

File: osu.Game/Screens/Play/BreaksOverlay/RemainingTimeCounter.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            isCounting = true;
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (isCounting)
+            {
+                var currentTime = Clock.CurrentTime;
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/RemainingTimeCounter.cs]: pretty sure this logic is wrong (the test case doesn't work as expected). `StartCounting` should probably accept and end time, not a remaining time.

File: osu.Game/Screens/Play/Player.cs [hunk 4]
```diff
@@ -161,21 +162,25 @@ private void load(AudioManager audio, OsuConfigManager config, APIAccess api)
...
                             Origin = Anchor.Centre
                         },
+                        new BreakOverlay(beatmap.BeatmapInfo.LetterboxInBreaks)
+                        {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                            Breaks = beatmap.Breaks,
+                            Clock = decoupledClock
+                        },
+                        new SkipButton(firstObjectTime) { AudioClock = decoupledClock },
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: Why did you change the height of the skip button?

File: osu.Game/Screens/Play/BreaksOverlay/LetterboxOverlay.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+
+namespace osu.Game.Screens.Play.BreaksOverlay
+{
+    public class LetterboxOverlay : Container
+    {
+        private const int letterbox_height = 350;
+
+        private Color4 transparentBlack => new Color4(0, 0, 0, 0);
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/LetterboxOverlay.cs]: should be a (static) field, not a property.

File: osu.Game/Screens/Play/BreaksOverlay/LetterboxOverlay.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Colour;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+
+namespace osu.Game.Screens.Play.BreaksOverlay
+{
+    public class LetterboxOverlay : Container
+    {
+        private const int letterbox_height = 350;
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/LetterboxOverlay.cs]: just `height` is fine

File: osu.Game/Screens/Play/BreaksOverlay/InfoContainer.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            Alpha = 0;
+            Direction = FillDirection.Vertical;
+            Spacing = new Vector2(5);
+            Children = new Drawable[]
+            {
+                new OsuSpriteText
+                {
+                    Anchor = Anchor.TopCentre,
+                    Origin = Anchor.TopCentre,
+                    Text = @"current progress".ToUpper(),
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/InfoContainer.cs]: should not have `@` prefix. this denotes non-localisable

File: osu.Game/Screens/Play/BreaksOverlay/InfoContainer.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                },
+                new FillFlowContainer<InfoLine>
+                {
+                    AutoSizeAxes = Axes.Both,
+                    Origin = Anchor.TopCentre,
+                    Anchor = Anchor.TopCentre,
+                    Direction = FillDirection.Vertical,
+                    Children = new []
+                    {
+                        new InfoLine(@"Accuracy", @"-"),
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/InfoContainer.cs]: English words should not have `@` prefix

File: osu.Game/Screens/Play/BreaksOverlay/GlowIcon.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+{
+    public class GlowIcon : Container
+    {
+        private const int blur_sigma = 5;
+
+        private readonly SpriteIcon spriteIcon;
+        private readonly SpriteIcon glowIcon;
+        private readonly BufferedContainer glowContainer;
+
+        public override Vector2 Size
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/GlowIcon.cs]: don't override size and give it a different meaning.

File: osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            if (letterboxing)
+                letterboxOverlay.FadeIn(fade_duration);
+
+            remainingTimeBox
+                .ResizeWidthTo(remaining_time_container_max_size, fade_duration, Easing.OutQuint)
+                .Then()
+                .ResizeWidthTo(0, b.Duration);
+
+            Scheduler.AddDelayed(() => remainingTimeCounter.StartCounting(b.EndTime), b.StartTime);
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs]: pretty sure this delay is wrong; you're not considering the current time.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/BreaksOverlay/GlowIcon.cs (hunk 9 above)
Comment (EVAST9919) [reply osu.Game/Screens/Play/BreaksOverlay/GlowIcon.cs]: Should i add my own property instead?

File: osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            InitializeBreaks();
+        }
+
+        public void InitializeBreaks()
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs]: This shouldn't be public, and should be automatically called on setting `Breaks`. Make sure when you make this change to clear all existing transforms (including scheduled ones), as there's currently an easily reproducible bug using your test cases (run the two-breaks testcase then any other; they will overlap and cause unexpected results).

**Comment by peppy** [comment]
I think the count should be one higher than it is right now (ie. it should never show 0).

File: osu.Game/Screens/Play/BreaksOverlay/ArrowsOverlay.cs [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics;
+using OpenTK;
+using osu.Game.Graphics.Containers;
+
+namespace osu.Game.Screens.Play.BreaksOverlay
+{
+    public class ArrowsOverlay : Container
```
Comment (huoyaoyuan) [inline osu.Game/Screens/Play/BreaksOverlay/ArrowsOverlay.cs]: Why not `OverlayContainer` (and other cases)?

File: osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+            if (letterboxing)
+                letterboxOverlay.FadeIn(fade_duration);
+
+            remainingTimeBox
+                .ResizeWidthTo(remaining_time_container_max_size, fade_duration, Easing.OutQuint)
+                .Then()
+                .ResizeWidthTo(0, b.Duration - fade_duration);
+
+            tasks.Add(new ScheduledDelegate(() => remainingTimeCounter.StartCounting(b.EndTime), b.StartTime));
+            Scheduler.Add(tasks.Last());
```
Comment (peppy) [inline osu.Game/Screens/Play/BreaksOverlay/BreakOverlay.cs]: doing a linq query just to avoid a single local variable? please no.

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs [hunk 14]
```diff
@@ -76,7 +76,7 @@ protected ScoreProcessor()
             Combo.ValueChanged += delegate { HighestCombo.Value = Math.Max(HighestCombo.Value, Combo.Value); };
         }
 
-        private ScoreRank rankFrom(double acc)
+        public static ScoreRank RankFrom(double acc)
```
Comment (peppy) [inline osu.Game/Rulesets/Scoring/ScoreProcessor.cs]: move this inside of score. we don't use static public methods in non-static classes

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
👍
